Syntax sugar for resolving overloaded function pointers

dean0x7d opened the pull request:
I'm actually hesitant about this PR because of the unusual syntax, so feel free to close this and label it "syntax gore" :) Anyway, here goes.

This tries to include something similar to [`overload_cast`](https://github.com/dean0x7d/overload_cast), but with a uniform C++11 and C++14 syntax. Overloaded functions need to be cast before passing to `def`:
```c++
.def("foo", static_cast<Return (Widget::*)(Arg0, Arg1, Arg2)>(&Widget::foo))
```
Here, the class name needs to be written twice, and the return type is completely unnecessary. The terser syntax introduced here is:
```c++
.def("foo", &Widget::foo | py::select<Arg0, Arg1, Arg2>())
```
`py::select` resolves the overload set just like the cast, but it only needs the arguments -- no extra `Return (Widget::*)()` noise. The `|` operator is an unfortunate evil here. Although, it can be interpreted as a shell-like pipe operator as is becoming popular in C++ (ranges TS). Ideally, `py::select` would just be an argument of `def` (so there would just be a normal comma there), but template type deduction can't resolve the function pointer in that case (at least with everything I've tried).

An alternative syntax is also possible with `overload_cast` as linked above, but this has its own syntax issues (split between C++11 and 14).

See the tests for examples. I can add proper docs if this feature is at all desired.

wjakob commented:
That's neat -- however, yet one more alternative would be to include the more elegant ``overload_cast`` syntax and simply make this a C++14 only API. This will become a non-issue shortly (actually, if it works on Visual studio, I'd say that it is basically already a non-issue).

dean0x7d commented:
Sure, that should work. (Making it C++14 exclusive totally slipped my mind -- I'm mostly stuck with C++11 because of Ubuntu 14.04 LTS and GCC 4.8.)

VS2015 Update 2 and newer can compile `overload_cast`. I'll need to see what the proper version check should be. 

I'll make the changes and add docs then.

wjakob commented:
Great, sounds good to me!

dean0x7d commented:
OK, switched to `overload_cast` with a guard for compatible compilers. Added an example to the docs. Let me know if that's OK.

Obligatory: Something, something, MSVC grumble.

wjakob commented:
Beautiful, thank you for the nice docs as well.
